Add sequence_nr to aot_autograd to map forward ops to their corresponding backward ops

alexmsettle opened the pull request:
Fixes #102375 

Sequence_nr increments in the forward pass and decrements in the backward pass.  Backward ops with the same sequence_nr as a forward op represent the backward implementation for the op.  The long term goal is to make this information available to the profiler so users can observe which ops are fused by the inductor openai triton kernels.

Added a test for this feature **test/dynamo/test_aot_autograd.py::AotAutogradFallbackTests::test_aot_sequence_nr**.  The test case uses **aot_export_module()** to create a joint fwd/bwd fx graph.  Then it walks all the nodes in fx graph using fx_graph.graph.nodes.   The seq_nr of each node is recorded in node.meta.  During the fwd pass the seq_nr increments and it decrements during the bwd pass.  This allows the user to map forward ops to their corresponding bwd ops which is useful for performance analysis.  

Expected output from the test case.

 SeqNr|OrigAten|SrcFn
0|aten.convolution.default|l__self___conv1
0|aten.add.Tensor|l__self___bn1
1|aten._native_batch_norm_legit_functional.default|l__self___bn1
2|aten.relu.default|l__self___relu1
3|aten.add.Tensor|add
4|aten.view.default|flatten
5|aten.t.default|l__self___fc1
6|aten.unsqueeze.default|l__self___fc1
7|aten.mm.default|l__self___fc1
8|aten.squeeze.dim|l__self___fc1
9|aten.add.Tensor|l__self___fc1
10|aten.sub.Tensor|l__self___loss_fn
11|aten.abs.default|l__self___loss_fn
12|aten.mean.default|l__self___loss_fn
12|aten.ones_like.default|
12|aten.expand.default|
12|aten.div.Scalar|
11|aten.sgn.default|
11|aten.mul.Tensor|
8|aten.unsqueeze.default|
7|aten.t.default|
7|aten.mm.default|
7|aten.t.default|
7|aten.t.default|
7|aten.mm.default|
6|aten.squeeze.dim|
5|aten.t.default|
4|aten.view.default|
2|aten.threshold_backward.default|
1|aten.native_batch_norm_backward.default|
0|aten.convolution_backward.default|
0|aten.add.Tensor|

      



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @anijain2305

alexmsettle committed:
(#102375) Added sequence_nr to aot_autograd to map

forward ops to their corresponding backward ops

alexmsettle committed:
Fix to AotConfig dataclass

alexmsettle committed:
Fixed lint errors

alexmsettle committed:
Added unit test

alexmsettle committed:
Added seq_id to list of preserved node meta fields

alexmsettle committed:
Fixed lint error

davidberard98 commented:
@alexmsettle is this PR ready for review or should I wait to review it?

alexmsettle committed:
Updated aot_sequence_nr test case to check for

seq_id in both the fwd and bwd fx graphs.

alexmsettle committed:
Removed print debugging statements

alexmsettle committed:
Added option to aot_export_module to preserve the

node meta data in the backward fx graph.

alexmsettle commented:
> @alexmsettle is this PR ready for review or should I wait to review it?

Hi @davidberard98  it is ready to review now.  I got stuck trying to make a test case that could read the fx.graph.nodes from the user level.  I think it is working now.

alexmsettle committed:
Fixed lint error

alexmsettle committed:
Fixed CI error related to device type

Fixed lint errors.

alexmsettle commented:
All the tests pass now so it is ready to review

davidberard98 commented on torch/_functorch/aot_autograd.py:
```diff
@@ -172,18 +174,22 @@ def prehook(grad_output):
 
             fx_traceback.set_stack_trace(stack_)
+            if stack_ != prev_stack and max_seq_id >= 0:
+                fx_traceback.set_bwd_seq_id(max_seq_id)
```
is it guaranteed that the backward ops will get called in exactly the reverse order compared to the forward ops? I'm not sure this is true...

I think it's also possible to find ops that will be decomposed in the dispatcher ("CompositeImplicitAutograd" ops), so that bwd_seq_id might become < -1 in some cases.

alexmsettle replied:
That's a good question, I'm not certain I can say guaranteed.  I think the idea is that the bwd pass is just walking the fwd graph in reverse order.  I think this is the assumption used in autograd but I'm not sure how to prove that it is always the case.

davidberard98 commented on torch/fx/traceback.py:
```diff
@@ -31,6 +32,20 @@ def set_stack_trace(stack : List[str]):
         current_meta["stack_trace"] = "".join(stack)
 
+@compatibility(is_backward_compatible=False)
+def set_bwd_seq_id(max_fwd_seq_id):
```
reminder to myself to check this again to make sure the globals are fine

alexmsettle replied:
I kept set_bwd_seq_id() but it no longer uses global variables.

davidberard98 reviewed (commented):
Hey, thanks for the PR.

- There's a few minor comments I left.
- There's also a potential issue with some assumptions (e.g. each fwd node has a corresponding bwd node, and they happen in a specific order) - for this I'll probably ask another reviewer (but he's on vacation this week, so I'll follow up next week)

alexmsettle commented on test/inductor/test_torchinductor.py:
```diff
@@ -2327,6 +2327,65 @@ def fn(x):
         )
         self.assertEqual(torch._inductor.metrics.generated_kernel_count, 0)
 
+    def test_aot_sequence_nr(self):
```
Moved to test/dynamo/test_aot_autograd.py

alexmsettle commented on test/inductor/test_torchinductor.py:
```diff
@@ -2327,6 +2327,65 @@ def fn(x):
+        print(f"Type of output {type(output)}")
+        expl, out_guards, graphs, _, _, verbose = torch._dynamo.explain(aot_fn, x, target)
+        for idx, graph in enumerate(graphs):
+            print(f"Graph {idx} listing {graph}")
```
This was resolved in an earlier commit.

alexmsettle commented on torch/_dynamo/symbolic_convert.py:
```diff
@@ -1890,6 +1890,7 @@ def __init__(
         # The first field of tuple is the fully qualified name of current module
         # in original hierarchy.  The second field is the type of current nn.module
         self.nn_module_stack: Dict[str, Tuple[str, Type[Any]]] = {}
+        self.fwd_seq_id = -1
```
Here's the reference to sequence_nr in autograd.  It looks like the sequence id is thread local, I didn't add any thread specific logic to account for this yet.   https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/function.h#L104-L111

alexmsettle replied:
I added a comment and changed the variable name to seq_id since it can refer to both fwd and bwd ops.  I also added a See Note [...] comment

alexmsettle commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3542,6 +3550,8 @@ def returned_function(*args, **kwargs):
         if cached_res is None:
             flat_fn, out_spec = create_tree_flattened_fn(fn, args, kwargs)
 
+            print(f"Calling create_aot_dispatch fn {fn}- no max fwd_seq_id set")
```
done

alexmsettle replied:
Found a couple others I forgot to remove earlier too.

alexmsettle committed:
Code review changes

alexmsettle commented:
Thanks for the review @davidberard98 .  I think I covered most of the issues you raised.  There are some more significant issues like is it guaranteed that the bwd ops are always being executed in reverse order of the fwd ops.  I'll try to come up with some better test cases.  Looks like I also need to make the seq_id thread local to be consistent with sequence_nr in autograd.

alexmsettle committed:
Fixed merge conflicts from main branch

Merge remote-tracking branch 'origin/master' into alex_aot_seq_id

Conflicts:
	test/dynamo/test_aot_autograd.py
	test/inductor/test_torchinductor.py
	torch/_functorch/aot_autograd.py
	torch/fx/proxy.py

alexmsettle committed:
Added missing import statements for new test case

davidberard98 reviewed (commented):
Hey, here are some suggestions from a discussion I had with @bdhirsh today:

1. Here's an alternative method of attaching sequence numbers: sequence number already exists in eager (non-dynamo, non-aot-autograd) autograd, so you might be able to just copy the autograd sequence number. You can grep for "sequence_nr" in the codebase for some examples (I see you've already touched this in the profiler before). You'll probably have to add python bindings in order to be able to access this information. Feel free to ask if you want more pointers or if you think there's a reason it might not work.
2. some examples of things that might fail the current tests:

* If the function has some nodes that don't run autograd (e.g. they don't have any inputs that require_grad) and other nodes that do run autograd. Then I would suspect that the forward method could have more nodes with a labeled sequence_nr than the backward graph will have, or it might have non-contiguous sequence_nrs
* If the function has ops like addmm, which will get decomposed

alexmsettle committed:
Fixed CI errors

alexmsettle commented:
I made this a draft PR while I consider how to implement the suggestions above.

alexmsettle committed:
Merge branch 'pytorch:main' into alex_aot_seq_id

alexmsettle committed:
Added functions to autograd to allow python

calls to get the sequence number.  The forward pass
uses the current value of the sequence number counter
get_sequence_nr().  For the bwd pass, each node saves its
sequence number as it is executed.  This requires another
function that can access the sequence id from the Node data
structure.

alexmsettle committed:
Updated aot test case with some debug statements.

These are temporary, will remove before merge to main.

alexmsettle commented:
> Hey, here are some suggestions from a discussion I had with @bdhirsh today:
> 
> 1. Here's an alternative method of attaching sequence numbers: sequence number already exists in eager (non-dynamo, non-aot-autograd) autograd, so you might be able to just copy the autograd sequence number. You can grep for "sequence_nr" in the codebase for some examples (I see you've already touched this in the profiler before). You'll probably have to add python bindings in order to be able to access this information. Feel free to ask if you want more pointers or if you think there's a reason it might not work.
> 2. some examples of things that might fail the current tests:
> 
> * If the function has some nodes that don't run autograd (e.g. they don't have any inputs that require_grad) and other nodes that do run autograd. Then I would suspect that the forward method could have more nodes with a labeled sequence_nr than the backward graph will have, or it might have non-contiguous sequence_nrs
> * If the function has ops like addmm, which will get decomposed

Hi @davidberard98 @bdhirsh thanks for the suggestion about using autograd directly to reference the sequence number.  I implemented helper functions to export the sequence_nr from cpp to pytorch.  It seems to work pretty well.  I have a little more debugging to complete but I think this approach is much better than before.

alexmsettle committed:
Fixed lint errors

alexmsettle committed:
Moved where the fwd seq id is set.

It happens right when the current node meta is created
in the traceback

alexmsettle committed:
Fixed lint errors.  Removed print debug statements

alexmsettle committed:
Fixed linterror in test case

alexmsettle committed:
Added sequence_nr method to torch.autograd.Function

so custom ops will have access to the sequence_nr.

alexmsettle committed:
lintrunner fix

alexmsettle committed:
Merge branch 'pytorch:main' into alex_aot_seq_id

soulitzer commented on torch/fx/traceback.py:
```diff
@@ -55,6 +67,8 @@ def set_current_meta(meta : Dict[str, Any]):
         try:
             current_meta = meta
+            # Record sequence ID right before the node is executed
+            current_meta["seq_id"] = get_sequence_nr()
```
This doesn't feel like the right granularity because this entrypoint interposes at the `call_{function},{module},{method}` level, so you are essentially capturing the sequence_nr corresponding to the first node created during that `call_function` - or is that the intended behavior? If it is the intended behavior, how do you plan to use this information downstream?

alexmsettle replied:
Hi @soulitzer thanks for the code review!  I see what you mean about where to set the seq_nr.  The common case is that setting the seq_nr at the call_{func},{mod},{method} level is sufficient. eg when the ops get lowered to aten they would maintain the same seq_nr.  However,  I found an example where this doesn't hold up.  See the attached profile clip from running in eager mode.  In this case l1_loss has seq_nr 28, but it gets lowered to **aten::abs seq=29** and **aten::mean seq=30**.  Then in the bwd pass **seq=29 and seq=30** are the nodes that have gradients computed and seq=28 doesn't appear in the bwd pass. 

![image](https://github.com/pytorch/pytorch/assets/37422826/4d06efc3-04af-4746-bbb6-92792d67aaad)

alexmsettle replied:
Ok I added some logic that does a better job at tracking the seq_nr for aten ops that are composed of other aten ops.  In the example above I think aten::l1_loss is treated as a **call_module** but this op is implemented by other aten ops eg aten::abs and aten::mean which are considered inner_ops by the dispatcher logic in torch fx.  I added logic to torch.fx.proxy.create_node() which detects that the op is in the fwd pass and adds updates the seq_nr with the most recent seq_nr captured by autograd.  

Here's an example of how you can use seq_nr for analysis.  I walked the joint fx graph and printed node and node.meta attributes to make this table.  You can see that **seq_nr 755** is the main **matmul** op for the linear layer.  In the bwd pass there are 2 more **aten::mm** ops and 3 **transpose** ops associated with 755.  These correspond to the data and weight gradient computations.  You can also see there is a transpose and squeeze (seq_nr 754 and 755) associated with this same linear op.  They each have a corresponding bwd op you can see at the bottom of the table.

This information is really useful for performance analysis.  The longer term goal with adding sequence_nr to aot_autograd is that I want to be able to observe this in the open ai triton kernels using the profiler.  This information will enable a user to better understand which operators from the original network graph were fused together in the triton kernels.
 
![aot_seq_nr](https://github.com/pytorch/pytorch/assets/37422826/1dbf9ef0-b449-4f77-becf-7f9d70012c80)

soulitzer commented on test/dynamo/test_aot_autograd.py:
```diff
@@ -720,6 +722,76 @@ def fn(x):
+                trace_joint=True,
+                output_loss_index=0,
+                skip_flatten_joint=True,
+            )
```
Could you add to the PR description what the output of this is, i.e. the joint graph + seq_nr information?

alexmsettle replied:
I just updated the PR description, is that sufficient?

soulitzer replied:
Thanks for the additional context there- I was also thinking we could have in the description a example of a traced joint graph, with all the nodes in the forward and backward as well as what the seq_nr in their metas are.

alexmsettle replied:
Yes - I just added it to the test, thanks for showing me how.

soulitzer commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -1361,6 +1368,7 @@ static struct PyGetSetDef THPFunction_properties[] = {
 // NOLINTNEXTLINE(modernize-avoid-c-arrays,cppcoreguidelines-avoid-c-arrays,cppcoreguidelines-avoid-non-const-global-variables)
 static struct PyMethodDef THPFunction_methods[] = {
     {(char*)"name", THPFunction_name, METH_NOARGS, nullptr},
+    {(char*)"sequence_nr", THPFunction_sequence_nr, METH_NOARGS, nullptr},
```
We probably want these to be private API for now, i.e. we should expose it as `_sequence_nr` as what it returns is not super straightforward for users

alexmsettle replied:
Makes sense - I changed it to _sequence_nr in both python_function.cpp and python_cpp_function.h

alexmsettle committed:
Removed unused code that I forgot to remove after

changing the way sequence_nr is obtained.

alexmsettle commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3675,6 +3682,7 @@ def aot_module_simplified(
     # Next, the input args
     full_args.extend(args)
 
+    mod_count = 0
```
Oh looks like I forgot to remove some code that is no longer needed.  I made a significant change to the way I was capturing the sequence_nr.  There are a few more examples of dead code I'm removing.

alexmsettle commented on torch/_functorch/aot_autograd.py:
```diff
@@ -114,7 +114,7 @@ def partial_asdict(obj: Any) -> Any:
 callback_set = False
 
-def setup_stacktrace_preservation_hooks(roots: List):
+def setup_stacktrace_preservation_hooks(roots: List, max_seq_id: int = 0):
```
Removed - same comment as above.  I changed the logic and forgot to clean this up.

alexmsettle commented on torch/fx/proxy.py:
```diff
@@ -138,7 +138,7 @@ def create_node(self, kind : str, target : Target,
             # Explicitly set the stack_trace, nn_module_stack and source_fn on the node.meta
             # If other meta fields are needed, they can be added here
-            copy_meta_fields = ["nn_module_stack", "source_fn", "original_aten", "recompute"]
+            copy_meta_fields = ["nn_module_stack", "source_fn", "original_aten", "recompute", "seq_id"]
```
Good point, I just changed all instances of seq_id to seq_nr.

alexmsettle commented on torch/_functorch/aot_autograd.py:
```diff
@@ -154,18 +154,21 @@ def prehook(grad_output):
     for node in iter_graph(roots):
         forward_node_stack = node.metadata.get("traceback_", [])
-        node.register_prehook(get_prehook(forward_node_stack))
+        forward_node_module_info = str(node.name)
```
Removed it.

alexmsettle commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3791,6 +3804,13 @@ def aot_export_module(
+    if isinstance(mod, torch.fx.GraphModule):
+        for node in mod.graph.nodes:
+            if node.op in ("call_function", "call_method", "call_module"):
+                max_fwd_ops += 1
```
Removed - more code I forgot to remove after changing the approach to getting the sequence_nr.

alexmsettle committed:
Code review changes

Removed dead code in _functorch/aot_autograd.py
Changed all instances of seq_id with seq_nr
Made Node::sequence_nr a private variable by changing the name to
Node::_sequence_nr.
Updated test to use seq_nr.

alexmsettle committed:
Resolved merge conflicts

Merge remote-tracking branch 'origin/master' into alex_aot_seq_id

Conflicts:
	test/dynamo/test_aot_autograd.py

alexmsettle committed:
For the forward pass, added nested more fine grain

sequence nr tracking for the case where a call_module op
is composed of multiple call_function aten ops.  Each of
which has a unique sequence_nr.

alexmsettle committed:
Fixed BC error in proxy.py

Fixed lint error in test_aot_autograd.py

alexmsettle commented:
Hi @soulitzer @davidberard98 could you please review my recent changes?  I think I have addressed the issues you raised.

alexmsettle committed:
Merge branch 'pytorch:main' into alex_aot_seq_id

alexmsettle committed:
Code review changes

alexmsettle committed:
Fixed lint error

alexmsettle reviewed (commented):
Thanks for the reviews @soulitzer . I added the expected output to the PR message and to the test case.

soulitzer commented on torch/fx/traceback.py:
```diff
@@ -32,6 +33,17 @@ def set_stack_trace(stack : List[str]):
+        # The seq_nr of the backward ops are saved
+        # in the autograd::Node data structure at the time
+        # the Node is created
+        current_meta["seq_nr"] = seq_nr
```
> Then create_node() is called and the contents of current_meta are copied to the new node which represents the backward op.

That makes sense, thanks!

soulitzer commented on torch/fx/interpreter.py:
```diff
@@ -189,6 +190,8 @@ def run_node(self, n : Node) -> Any:
         """
         with self._set_current_node(n):
+            # Sets the seq_nr for the fwd pass
+            fx_traceback.set_seq_nr(get_sequence_nr())
```
Still not sure why this is needed. It doesn't look like we we check what the value of this is later. And if during the forward pass it would be overwritten by new_seq, and if during the backward pass it would be overwritten by whatever the hooks set.

alexmsettle replied:
Removed it.

soulitzer reviewed (commented):
Thanks for the update, looks mostly good. I added some comments.

The main feedback I have though is that adding a new flag does not seem ideal, I'm wondering if there's a simple way to avoid this. For example, while retracing, run fx_g under an interpreter? e.g. something like ` Interpreter(fx_g).run(*args, **kwargs)`.

That way the global node meta would be set, and we can propagate the node meta to the new graph traced by fx.

We should also be careful about sequence_nr continuing to increase if `create_graph=True`.

alexmsettle committed:
Fixed regression error

alexmsettle committed:
Changed test case to compare expected text output

Walk the fx graph, print the nodes, compare to expected output.

alexmsettle committed:
Code review changes

Made proxy.py the only place where node.meta["seq_nr"] is set.
Updated the test case to include expected output text.
Added "bwd_op" field to node.meta so we know how to apply
the sequence_nr to the node.

alexmsettle commented on torch/fx/proxy.py:
```diff
@@ -142,6 +142,20 @@ def create_node(self, kind : str, target : Target,
+                new_seq_nr = current_meta["seq_nr"]
+                current_meta["bwd_op"] = False
+            node.meta["seq_nr"] = new_seq_nr
+
```
I finally got setting node.meta["seq_nr"] isolated to 1 place.
I added a field to current_meta to call out bwd_ops.  Then in create_node() the seq_nr for bwd ops is taken from current_meta and for fwd ops it is taken from get_sequence_nr().

alexmsettle commented on torch/fx/traceback.py:
```diff
@@ -32,6 +32,19 @@ def set_stack_trace(stack : List[str]):
+        current_meta["seq_nr"] = seq_nr
+        if bwd:
+            current_meta["bwd_op"] = True
+
```
I added "bwd_op"  to help create_node() get the sequence_nr from the right place.

alexmsettle commented:
> obal node meta would be set, and we can propagate the node meta to the new graph traced by fx.

@soulitzer 
What if instead of making changes to **aot_export_module()** I just move the code I need from aot_export_module into my test case.  Really all I need to be able to do is call **_aot_export_function()**. That will give me the fx_graph I need to verify that sequence_nr is working.  I tried this locally and it works.

soulitzer commented:
> What if instead of making changes to aot_export_module() I just move the code I need from aot_export_module into my test case. Really all I need to be able to do is call _aot_export_function(). That will give me the fx_graph I need to verify that sequence_nr is working. I tried this locally and it works.

@alexmsettle I think the question is just do we actually want to provide a public API or not at this point?

In other words, for the applications we plan to use this for downstream (profilers, etc) would we be able to read the output of _aot_export_function() directly, or would we prefer to be able to read from aot_export_module()? 

I think @davidberard98 should weigh in on this,  but my understanding is that aot_export_module does some of the normalization of inputs and outputs that is important for downstream applications, so it is not immediately clear to me.

bdhirsh commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3742,6 +3744,10 @@ def aot_export_module(
+    # For debug purposes the user may want to inspect the joint graph.
+    # Skipping the flatten_joint stage perserves the node.meta information
+    # available in the joint fwd/bwd graph fx_g.
+    skip_flatten_joint: int = False,
```
Like a few others have mentioned - I'm not a huge fan of (yet) another config in AOTAutograd. Would it be possible to:

It's a bit unclear to me what purpose this serves:
(1) this flag will force you to only generate a forward graph? Why not just pass `trace_joint=False` in your example?
(2) more of general question: is important/useful to use the profiler with `torch.export`, which is what your testing in the the test case from `test_aotdispatch.py`? Or is the benefit mostly from working with `torch.compile` which doesn't seem like it needs any changes).

alexmsettle replied:
Thanks for the feedback @bdhirsh .  I went ahead and removed the changes I made to aot_export_module().  I originally introduced these changes because I needed a way to observe from code that the sequence_nr() was being assigned correctly to each op.  aot_export_module() allowed me to do this, but only if I forced the control flow to bypass a call to make_fx().  I found a way to make the test case cover this.  The intent of the sequence_nr() changes is that the profiler or developer can observe the sequence_nr() from downstream inductor (torch/_inductor/scheduler.py   origin_ops).

alexmsettle committed:
Merge branch 'pytorch:main' into alex_aot_seq_id

alexmsettle committed:
Removed changes made to aot_export_module

Updated test case to cover some of the functionality of
aot_export_module in the test.

alexmsettle commented on torch/_functorch/aot_autograd.py:
```diff
@@ -3851,6 +3855,13 @@ def fn_to_trace(*args):
+    # available in the joint fwd/bwd graph fx_g.
+    if skip_flatten_joint:
+        trace_joint = False
+
```
I went ahead and removed all the changes I made to aot_export_module.  I made the changes initially because I needed a way to write a test case for the sequence_nr() related changes.  I found a workaround so I only have to modify the test case.

alexmsettle committed:
Fixed failing test case.

Made sequence_nr() reporting start at 0 for the test case.

alexmsettle commented on torch/fx/traceback.py:
```diff
@@ -55,6 +67,8 @@ def set_current_meta(meta : Dict[str, Any]):
         try:
             current_meta = meta
+            # Record sequence ID right before the node is executed
+            current_meta["seq_nr"] = get_sequence_nr()
```
Done - I removed this code.  Now the node.meta["seq_nr"] is set in the create_node() method in proxy.py for both the fwd and the bwd pass.

alexmsettle committed:
Added new regression tests

Added support for autograd create_graph=True option when
generating the BWD graph.
Added thread_id to python c++ interface because seq_nr is thread
local.

alexmsettle committed:
Fixed conflict in proxy.py

Merge remote-tracking branch 'origin/master' into alex_aot_seq_id

Conflicts:
	torch/fx/proxy.py

alexmsettle committed:
Fixed lint errors

alexmsettle committed:
Fixed build error

alexmsettle committed:
Fixed CI test errors.

davidberard98 commented on torch/_C/_autograd.pyi:
```diff
@@ -90,6 +90,8 @@ def _disable_profiler() -> _ProfilerResult: ...
 def _profiler_enabled() -> bool: ...
 def _add_metadata_json(key: str, value: str) -> None: ...
 def _kineto_step() -> None: ...
+def get_sequence_nr() -> int: ...
```
very minor nit - these can probably be private as well. (you could leave these for a follow-up PR if you prefer)

alexmsettle replied:
Done - I changed the name to _get_sequence_nr().  I went ahead and removed get_thread_id(), it turns out it wasn't needed afterall.

davidberard98 reviewed (commented):
I think everything looks good (left a very minor comment) - I just imported to internal where there are some extra tests for FX. I'll check back later today or tomorrow morning - as long as those extra tests pass, I think this should be good to go!

alexmsettle commented:
> Thanks for the update, looks mostly good. I added some comments.
> 
> The main feedback I have though is that adding a new flag does not seem ideal, I'm wondering if there's a simple way to avoid this. For example, while retracing, run fx_g under an interpreter? e.g. something like ` Interpreter(fx_g).run(*args, **kwargs)`.
> 
> That way the global node meta would be set, and we can propagate the node meta to the new graph traced by fx.
> 
> We should also be careful about sequence_nr continuing to increase if `create_graph=True`.

Hi @soulitzer - I have been working on the `create_graph=True` case.  I don't know a good way to test this because create_graph=True is not currently an option for `aot_autograd.create_joint`.  However, I did try adding the option to my local copy of the code and it does not interfere with the way I'm accounting for the backward pass sequence_nr.  Eg the backward ops continue to decrement even though the main sequence_nr counter is also increasing.  Any suggestions for how to make a test that shows this case works as expected?

soulitzer reviewed (commented):
Thanks for the update! Sounds mostly good on the autograd side, though I'd expect it would be a private API used in a non-load-bearing way (as with the autograd node sequence_nr), since there's some edges like foreach, multithreading, etc that complicate things.

I'd agree with @davidberard98 that the new get_sequence_nr and get_thread_id should be made private.

soulitzer commented:
@alexmsettle  The double backward case is more complicated to support because that would mean that any particular node that we capture while running backward would need to hold two sequence_nrs. I doubt its super important that we support that right now. 
I think the updated method where you pass in is_bwd to set_seq_nr is definitely preferred that the previous method which seemed quite brittle. So thats already quite good, even if we don't support double backward today!

soulitzer commented on test/dynamo/test_aot_autograd.py:
```diff
@@ -715,6 +720,190 @@ def fn(x):
+        ) as kineto_prof:
+            res = model_instance(*args)
+            (gx,) = torch.autograd.grad(
+                res[0], x, create_graph=True, grad_outputs=grad_output
```
Leave a comment mentioning that we purposefully keep create_graph=True in order to test that the new sequence_nr increments don't interfere with the setting of metadata during backward graph tracing

alexmsettle replied:
Done

alexmsettle committed:
Merge branch 'pytorch:main' into alex_aot_seq_id

alexmsettle committed:
Made get_sequence_nr() and get_thread_id() private

alexmsettle committed:
Picking up latest changes from main torch repo.

Should fix a regression failure.

Merge remote-tracking branch 'origin/alex_aot_seq_id' into alex_aot_seq_id

alexmsettle committed:
Fixed a type in function cpp exported function.

Updated test case for create_graph true case.

alexmsettle committed:
Removed _get_thread_id export from autograd cpp.

It is not needed to record the correct sequence_nr in aot with
or without create_graph=True.

alexmsettle committed:
Merge branch 'pytorch:main' into alex_aot_seq_id

alexmsettle committed:
Fixed lint error

alexmsettle committed:
Testing locally with latest changes from remote

Merge remote-tracking branch 'origin/alex_aot_seq_id' into alex_aot_seq_id

alexmsettle commented on torch/fx/traceback.py:
```diff
@@ -32,6 +32,19 @@ def set_stack_trace(stack : List[str]):
+        # in the autograd::Node data structure
+        # while the network is being traced.
+        current_meta["seq_nr"] = seq_nr
+        current_meta["fwd_trace_complete"] = bwd
```
Done

alexmsettle commented on torch/fx/proxy.py:
```diff
@@ -142,6 +142,23 @@ def create_node(self, kind : str, target : Target,
+            new_seq_nr = torch.autograd.get_sequence_nr() - 1
+            thread_id = torch.autograd.get_thread_id()
+            # In the FWD pass the new_seq_nr number maintained by
+            # torch.autograd.get_sequence_nr() keeps increasing.
```
Done

alexmsettle commented on torch/fx/proxy.py:
```diff
@@ -142,6 +142,23 @@ def create_node(self, kind : str, target : Target,
+            # torch.autograd.get_sequence_nr() keeps increasing.
+            # During the bwd pass the seq_nr for bwd ops is maintained
+            # in the torch::autograd::Node class see NOTE [ Sequence Number ].
+            # The seq_nr decrements during the bwd pass.
```
Done

alexmsettle commented on torch/fx/proxy.py:
```diff
@@ -142,6 +142,23 @@ def create_node(self, kind : str, target : Target,
+            # The seq_nr decrements during the bwd pass.
+            saved_th_id = current_meta.get("thread_id", -1)
+            if current_meta.get("fwd_trace_complete", False) and\
+                    thread_id == saved_th_id:
```
I went ahead and removed the thread_id logic.  I thought I needed it to support the create_graph=True case, but it turns out it's not needed.

alexmsettle committed:
Cleaned up comments from code review

alexmsettle committed:
Fixed lint error

soulitzer reviewed (approved):
LGTM, thanks for the updates!

davidberard98 commented:
@pytorchbot merge
